src/main/java/org/codehaus/mojo/taglist/TagListReport.java

```diff
@@ -435,30 +453,33 @@ private List pruneSourceDirs( List sourceDirectories )
     }
 
     /**
-     * Checks whether the given directory contains Java files.
+     * Checks whether the given directory contains source files.
      * 
      * @param dir the source directory.
-     * @return true if the folder or one of its subfolders contains at least 1 Java file.
+     * @return true if the folder or one of its subfolders contains at least 1 source file that matches includes/excludes.
      */
     private boolean hasSources( File dir )
     {
         boolean found = false;
         if ( dir.exists() && dir.isDirectory() )
         {
-            File[] files = dir.listFiles();
-            for ( int i = 0; i < files.length && !found; i++ )
-            {
-                File currentFile = files[i];
-                if ( currentFile.isFile() && currentFile.getName().endsWith( ".java" ) )
-                {
+            try {
+                if ( ! FileUtils.getFiles( dir, getIncludesCommaSeparated(), getExcludesCommaSeparated() ).isEmpty() ) {
                     found = true;
                 }
-                else if ( currentFile.isDirectory() )
-                {
-                    boolean hasSources = hasSources( currentFile );
-                    if ( hasSources )
-                    {
-                        found = true;
+            } catch (IOException e) {
+                // should never get here
+                getLog().error("Error whilst trying to scan the directory " + dir.getAbsolutePath(), e);
+            }
+            File[] files = dir.listFiles();
+            if ( files != null ) {
+                for ( int i = 0; i < files.length && !found; i++ ) {
+                    File currentFile = files[i];
+                    if ( currentFile.isDirectory() ) {
+                        boolean hasSources = hasSources( currentFile );
+                        if ( hasSources ) {
+                            found = true;
```

Comment:
I think you can just 
```suggestion
                            return true;
```
here, because further advancing the directory tree will not give any different results.

Reply:
I initially did that but then thought that you would prefer as few changes as possible :-(

Reply:
Also, I don't understand why we check the directories first, it seems like overkill, without this we would just get the list of files from all the directories and do nothing if the list is empty?

Reply:
Both good points. While "return" is a small fix, the other one could go into a new issue (and into the 3.1.0 release). If you like, open a new issue. That would be most helpful! :)